### PR TITLE
feat(shim): forward ro bind-mount option to virtio-fs share

### PR DESCRIPTION
**Pull request by ndeloof**
## Summary

- Parse the `ro` OCI mount option in `bindMounter.FromBundle` and carry it on `bindMount`, mirroring what the ext4 branch in the same file already does.
- Forward the flag through `sandbox.WithFS` to `vmInstance.AddFS`, which now applies `vm.MountOpts` before calling into libkrun.
- Bind the optional `krun_add_virtiofs3` symbol (libkrun >= 1.18) via a small `registerOptionalLibFunc` helper and a `,optional` struct-tag flag; route `AddVirtiofs` through it when present so the share is marked read-only at the host edge.
- Fall back to `krun_add_virtiofs` with a single warn log when the v3 symbol is not exported; guest-side `ro` enforcement via OCI mount options is unchanged.
- Cover the new behavior with unit tests for `parseCTag` and the `AddVirtiofs` v3-vs-legacy matrix (prefers v3, falls back when v3 is nil, surfaces non-zero return codes on both paths, errors when neither symbol is bound).

## Behavior

| libkrun version | Effect of this change |
| --- | --- |
| `< 1.18` (no `krun_add_virtiofs3` exported, e.g. the v1.17.4 currently pinned) | No-op on the host side. The fallback path runs and logs a single warn for read-only shares; the guest continues to enforce `ro` through the OCI mount options exactly as today. |
| `>= 1.18` (`krun_add_virtiofs3` exported) | The virtio-fs share is marked read-only at the host edge, so the VMM rejects writes in addition to the guest bind mount. This is defense-in-depth and lets the VMM treat the share as immutable for any future caching optimizations. |

## Test plan

- [ ] `go build ./...`
- [ ] `go test ./internal/shim/task/... ./internal/vm/libkrun/...`

### CODE REVIEW

File: internal/vm/libkrun/krun.go [hunk 1]
```diff
@@ -335,10 +359,32 @@ func openLibkrun(path string) (_ *libkrun, _ uintptr, retErr error) {
...
 	for i := 0; i < ik.NumField(); i++ {
-		cName := ik.Type().Field(i).Tag.Get("C")
+		tag := ik.Type().Field(i).Tag.Get("C")
+		cName, optional := parseCTag(tag)
 		fn := ik.Field(i).Addr().Interface()
+		if optional {
+			// registerOptionalLibFunc reports whether the symbol was
+			// resolved; if not, the field is left nil and call sites
+			// must guard against it.
+			registerOptionalLibFunc(fn, f, cName)
```
[the remark this answers is not in the document]
Comment (ndeloof) [reply internal/vm/libkrun/krun.go]: Addressed in 2bb72469bc191468cbf51bf8ba6c64a255a7eea6. The loader now logs a debug line via `log.L` when an optional libkrun symbol is not resolved, so operators can diagnose unexpected fallbacks (e.g. older libkrun missing `krun_add_virtiofs3`) by raising the shim log level.

File: internal/vm/libkrun/krun_test.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+	cases := []struct {
+		name     string
+		in       string
+		wantName string
+		wantOpt  bool
+	}{
+		{"plain", "krun_create_ctx", "krun_create_ctx", false},
+		{"optional", "krun_add_virtiofs3,optional", "krun_add_virtiofs3", true},
+		{"optional_with_unknown_tail", "krun_foo,optional,bogus", "krun_foo", true},
+		{"unknown_tail_only", "krun_foo,bogus", "krun_foo", false},
```
[the remark this answers is not in the document]
Comment (ndeloof) [reply internal/vm/libkrun/krun_test.go]: Fixed in 2bb72469bc191468cbf51bf8ba6c64a255a7eea6. `parseCTag` now returns an error on unknown tokens and the loader propagates it, so typos in `C:"..."` tags will fail fast at `openLibkrun` rather than silently leaving a function pointer nil. Tests updated accordingly.

File: internal/vm/libkrun/krun.go [hunk 3]
```diff
@@ -146,13 +149,32 @@ func (vmc *vmcontext) AddVSockPortConnect(port uint32, path string) error {
...
+	}
 	if vmc.lib.AddVirtiofs == nil {
 		return fmt.Errorf("libkrun not loaded")
 	}
+	if readonly {
+		log.G(ctx).WithField("tag", tag).Info(
+			"libkrun does not export krun_add_virtiofs3; " +
+				"read-only virtiofs enforcement deferred to the guest mount options",
+		)
+	}
```
[the remark this answers is not in the document]
Comment (ndeloof) [reply internal/vm/libkrun/krun.go]: Fixed in 2bb72469bc191468cbf51bf8ba6c64a255a7eea6. The fallback notice is now gated by a package-level `sync.Once` (the absence of `krun_add_virtiofs3` is a process-wide capability of the loaded library, so once is sufficient). No more per-mount log noise on hosts with older libkrun.

File: internal/vm/libkrun/krun.go [hunk 4]
```diff
@@ -261,6 +283,7 @@ type libkrun struct {
 	StartEnter         func(ctxID uint32) int32                                                               `C:"krun_start_enter"`
 	AddVsockPort       func(ctxID, port uint32, path string, listen bool) int32                               `C:"krun_add_vsock_port2"`
 	AddVirtiofs        func(ctxID uint32, tag, path string) int32                                             `C:"krun_add_virtiofs"`
```
[the remark this answers is not in the document]
Comment (ndeloof) [reply internal/vm/libkrun/krun.go]: Addressed in 2bb72469bc191468cbf51bf8ba6c64a255a7eea6 with a doc comment on the `AddVirtiofs3` field clarifying that `shmSize=0` means "use the libkrun default" (DAX window size), and noting that we can plumb a caller value through `MountConfig` when an actual use case appears. Keeping the API surface minimal for now since no caller currently needs a non-default value.

**Comment by dmcgowan** [comment]
> Fall back to krun_add_virtiofs with a single warn log when the v3 symbol is not exported; guest-side ro enforcement via OCI mount options is unchanged.

I don't think we need the fallback, we should just build libkrun at a specific version and support that as the minimum. The prepackaged builds have not been reliable.

File: internal/vm/libkrun/krun.go [hunk 5]
```diff
@@ -146,13 +150,42 @@ func (vmc *vmcontext) AddVSockPortConnect(port uint32, path string) error {
...
+// libkrun lacks krun_add_virtiofs3 and a read-only share is requested. The
+// fallback is a process-wide capability of the loaded library, so logging
+// it once per process is sufficient and avoids a steady stream of
+// identical messages on every container start.
+var virtiofsFallbackWarnOnce sync.Once
+
+// AddVirtiofs attaches a virtio-fs share to the VM. When readonly is true
+// it prefers krun_add_virtiofs3 (libkrun >= 1.18) which natively marks the
+// share read-only at the host edge. If that symbol is not exported by the
+// loaded libkrun, it falls back to krun_add_virtiofs and logs an info
```
Comment (akerouanton) [inline internal/vm/libkrun/krun.go]: As mentioned by Derek, we can just set the minimum version to 1.18 and remove all the fallback code here.

File: internal/shim/task/mount.go [hunk 6]
```diff
@@ -238,10 +243,29 @@ func (bm *bindMounter) FromBundle(ctx context.Context, b *bundle.Bundle) error {
 			specSrc = path.Join(vmTarget, filepath.Base(m.Source))
 		}
 
+		// Honor a read-only request from the OCI spec by marking the
+		// virtiofs share read-only at the host edge. The guest still
+		// applies the `ro` mount option to enforce read-only inside the
+		// container; this flag is defense in depth and ensures the host
+		// also rejects writes when the underlying libkrun supports it.
```
Comment (akerouanton) [inline internal/shim/task/mount.go]: > when the underlying libkrun supports it

Since we're bumping the minimum libkrun version, it always support `krun_add_virtiofs3`. Also, the defense-in-depth mechanism is now the guest-side `ro` option as it's correctly enforced on the host.

**Review by akerouanton — COMMENTED** [review]
Left a nit, otherwise LGTM.

**Review by akerouanton — APPROVED** [review]
LGTM

**Review by dmcgowan — APPROVED** [review]
